### PR TITLE
feat: add SES experiment toggle (iOS)

### PR DESCRIPTION
**Pull request by leotm**
## **Description**

A feature toggle for SES, within our settings menu under experiments, enabled by default

TODO
- [x] add `react-native-mmkv`
- [x] MVP
- [x] ensure only visible on iOS not Android
  - show only on iOS
  - or show if storage key exists
- [x] check next to Blockaid experiment enabled
- [x] fix test, update snapshot
- [x] add descriptive content
- [x] translate text
- [x] update styling
- [x] <s>replace `react-native-mmkv` with `react-native-default-preference`</s>
  - attempted in https://github.com/MetaMask/metamask-mobile/pull/8373/commits/ea5674c5600d5d11d78a07ff4ab774cf3848989f but only [partially working](https://github.com/MetaMask/metamask-mobile/pull/8373#issuecomment-1908544768)
- [x] fix Android debug bundling on Hermes
- [x] <s>consider content revision to include _Hardened JS_  and _SES (Secure EcmaScript)_ terms</s>
- [x] update to newly discussed text with link
- [x] fix styling, move switch above next to title
- [x] display both SES and Blockaid under same _Security_ heading
- [x] preserve Android UI
- [x] update SES to be enabled by default

## **Related issues**

Fixes: partially mitigates rollout risk

Follow-up to
- https://github.com/MetaMask/metamask-mobile/pull/8033

## **Manual testing steps**

iOS
- boot app
- login / create wallet
- go to bottom tab Settings menu
- open Experimental menu
- _SES enabled by default, like currently in (main) branch_
- disable SES ft toggle
- reboot app
- _SES now disabled_
- verify in Experimental menu
- repeat steps above to re-enable

Android
- boot app
- login / create wallet
- go to bottom tab Settings menu
- open Experimental menu
- UI remains the same
  - Blockaid shown if feature is enabled

## **Screenshots/Recordings**

Previous MVP video (old UI) [demo](https://consensys.slack.com/archives/C02U025CVU4/p1706035121992709?thread_ts=1705527596.738189&cid=C02U025CVU4) showing the behaviour, a simple settings menu ft toggle that persists, enabling/disabling lockdown upon app reboot, on iOS only

Android remains unchanged ([Blockaid disabled](https://github.com/MetaMask/metamask-mobile/assets/1881059/d74a30be-e65e-4832-a21e-8aa1812df608), [Blockaid enabled](https://github.com/MetaMask/metamask-mobile/assets/1881059/77b9a24b-486b-4e6b-aa10-51eede8c2951))

### **Before**

Blockaid ft disabled

Screenshot 2024-01-26 at 7 20 46 pm

Blockaid ft enabled

Screenshot 2024-01-26 at 7 25 05 pm

### **After**

Blockaid ft disabled

Screenshot 2024-01-30 at 1 07 07 pm

Blockaid ft enabled

Screenshot 2024-01-30 at 1 07 48 pm

Navigating to [Learn more](https://github.com/endojs/endo/blob/master/packages/ses/README.md)

Screenshot 2024-01-26 at 6 52 17 pm

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: patches/react-native+0.71.15.patch [hunk 1]
```diff
@@ -1,16 +1,18 @@
...
+index 1379ffd..4763ab2 100644
 --- a/node_modules/react-native/Libraries/Core/InitializeCore.js
 +++ b/node_modules/react-native/Libraries/Core/InitializeCore.js
-@@ -24,26 +24,53 @@
+@@ -24,26 +24,55 @@
  
  'use strict';
  
-+const Platform = require('../Utilities/Platform');
++import { MMKV } from 'react-native-mmkv';
```
Comment (leotm) [inline patches/react-native+0.71.15.patch:9]: no luck with `require` here

Screenshot 2024-01-23 at 5 15 58 pm
Screenshot 2024-01-23 at 5 16 07 pm

 but `import` works

**Comment by sethkfman** [comment]
@gauthierpetetin @Cal-L Can you take a look at this contribution to make sure it aligns with our platform for a product perspective?

@leotm Is there a more descriptive Header and content we can add so users and understand? Maybe something like Lockdown - This feature prevents dynamic modification of JavaScript running in the application.

**Comment by gauthierpetetin** [comment]
Hi @sethkfman , sorry I only see this now. I think end users won't understand the feature in its current form. I'll setup a meeting with @yanrong-chen @coreyjanssen @leotm and @hesterbruikman to discuss how we could present it.

**Comment by leotm** [comment]
@sethkfman i refactored to replace mmkv with default-preference https://github.com/MetaMask/metamask-mobile/pull/8373/commits/ea5674c5600d5d11d78a07ff4ab774cf3848989f
but had to revert to mmkv, since default-preference only partially works here as a toggle
(fine in settings menu, but in InitializeCore [.get](https://github.com/MetaMask/metamask-mobile/pull/8373/commits/ea5674c5600d5d11d78a07ff4ab774cf3848989f#diff-982477f8c8bf57eb44d53f104ddf22fb69df167560a37ad981551317419bb46bR13) is unable to return the storage value)

File: app/components/Views/Settings/ExperimentalSettings/index.tsx [hunk 2]
```diff
@@ -164,9 +176,47 @@ const ExperimentalSettings = ({ navigation, route }: Props) => {
...
+          />
+        </View>
+      </View>
+    </>
+  );
+
   return (
     <ScrollView style={styles.wrapper}>
       <WalletConnectSettings />
+      {Platform.OS === 'ios' && <SesSettings />}
```
Comment (jpcloureiro) [inline app/components/Views/Settings/ExperimentalSettings/index.tsx:219]: We can use `isIos` boolean from `./app/util/device` instead

https://github.com/MetaMask/metamask-mobile/blob/2b66c53abbcf63440c179e8bebf64b0033c804d1/app/util/device/index.js#L15-L17

File: patches/react-native+0.71.15.patch (hunk 1 above)
Comment (leotm) [inline patches/react-native+0.71.15.patch:9]: mmkv import breaking Hermes debug bundle on Android archs
splitting toggle logic to file required on iOS only should be sufficient
due to mmkv's async functions/generators unsupported on Hermes

File: app/components/Views/Settings/ExperimentalSettings/index.tsx (hunk 2 above)
Comment (leotm) [reply app/components/Views/Settings/ExperimentalSettings/index.tsx]: updated ^^

**Comment by gauthierpetetin** [comment]
Suggestion agreed with @coreyjanssen @leotm and @hesterbruikman :

- Title: Code lockdown
- Description: This is a test feature that blocks any changes to the app's javascript code without permission. [Learn more](https://github.com/endojs/endo/blob/master/packages/ses/README.md).

**Comment by hesterbruikman** [comment]
@naugtur IMO we should not release SES without a toggle to disable. My understanding is that conventional dapp use will not break, but we can't possibly test for all dapp use cases and so users need to be able to turn this feature off.

Is there a reason this PR is in draft?

If SES is already in 7.16.0 RC I imagine we want to add merge and add this PR asap.

- Has any `Learn more` copy been prepared? @naugtur please let me know so we can involve the documentation team to add knowledge base content to link out to.
- Believe this needs a translation PR for the copy

**Comment by leotm** [comment]
> @naugtur IMO we should not release SES without a toggle to disable. My understanding is that conventional dapp use will not break, but we can't possibly test for all dapp use cases and so users need to be able to turn this feature off.
> 
> Is there a reason this PR is in draft?
> 
> If SES is already in 7.16.0 RC I imagine we want to add merge and add this PR asap.
> 
> * Has any `Learn more` copy been prepared? @naugtur please let me know so we can involve the documentation team to add knowledge base content to link out to.
> * Believe this needs a translation PR for the copy

agreed been pushing to get this merged last week prior our Thurs RC cut (extended to Fri)

so presuming this will need cherry-pick into 7.16.0

i converted this PR to draft only now, only to make the change of SES enabled by default instead of disabled
then updating the screenshots and re-readying up
(i think this is more convenient than merging this, then 2 PRs to be cherry-picked)

> Suggestion agreed with @coreyjanssen @leotm and @hesterbruikman :
> 
> * Title: Code lockdown
> * Description: This is a test feature that blocks any changes to the app's javascript code without permission. [Learn more](https://github.com/endojs/endo/blob/master/packages/ses/README.md).

last Fri we agreed the Learn more link to be this above, who should i speak to clarify this is ok?

**Comment by hesterbruikman** [comment]
Thanks @leotm ! And apologies, hadn't noticed the link for Learn more. Can you please explain the trade-off in enabling this feature?

I'd like to understand if we can treat the toggle as a temporary Experimental fallback, in case of unforeseen issues. Or if long-term there might always be end-user use cases to disable the feature under Advanced. In case of the latter we can reach out to the documentation team to create Knowledge Base content.

**Comment by leotm** [comment]
> Thanks @leotm ! And apologies, hadn't noticed the link for Learn more. Can you please explain the trade-off in enabling this feature?

np! if by feature we mean this toggle for SES, the trade-off: our users might disable SES (reduce security), in the hopes of fixing an edge case of functionality (not yet covered by our tests)

or if by feature we mean SES, the trade-off: better security, at the compromise of possibly encountering an edge case of functionality not working as expected (i.e. discovery of a library we're using that calls non-standard JS)

> I'd like to understand if we can treat the toggle as a temporary Experimental fallback, in case of unforeseen issues. Or if long-term there might always be end-user use cases to disable the feature under Advanced. In case of the latter we can reach out to the documentation team to create Knowledge Base content.

exactly the toggle is a temp Experimental fallback, long-term we want SES enabled indefinitely (like in mm-extension) for better security (permanently prevent prototype pollution - now that's a tongue twister)

**Review by sethkfman — DISMISSED** [review]
LGTM

**Comment by hesterbruikman** [comment]
> > Thanks @leotm ! And apologies, hadn't noticed the link for Learn more. Can you please explain the trade-off in enabling this feature?
> 
> np! if by feature we mean this toggle for SES, the trade-off: our users might disable SES (reduce security), in the hopes of fixing an edge case of functionality (not yet covered by our tests)
> 
> or if by feature we mean SES, the trade-off: better security, at the compromise of possibly encountering an edge case of functionality not working as expected (i.e. discovery of a library we're using that calls non-standard JS)
> 
> > I'd like to understand if we can treat the toggle as a temporary Experimental fallback, in case of unforeseen issues. Or if long-term there might always be end-user use cases to disable the feature under Advanced. In case of the latter we can reach out to the documentation team to create Knowledge Base content.
> 
> exactly the toggle is a temp Experimental fallback, long-term we want SES enabled indefinitely (like in mm-extension) for better security (permanently prevent prototype pollution - now that's a tongue twister), then ofc run LavaMoat (which [runs](https://github.com/LavaMoat/LavaMoat/blob/main/packages/core/src/kernelTemplate.js#L54-L67) on SES)

Just to conclude on this: I don't think we need further `Learn more` end-user documentation.

**Review by sethkfman — APPROVED** [review]

**Pull request merged**